### ISSUE TITLE
executeCodeInNs?

### ISSUE THREAD

**Issue by qwtel** [jasongilman/proto-repl#10]
I was investigating why the repl would always print `user=>`, no matter what namespace I was in. 

Example:

``` clojure
(defn foo [x] (* x x))
; #'user/foo
; user=>

(foo 3)
; 9
; user=> 

(ns my.ns)
; nil
; user=>

(foo 3)
; CompilerException java.lang.RuntimeException: Unable to resolve symbol: foo in this context, compiling: ...
; user=>
```

Doing the same in a `lein repl` would print `my.ns=>` after sending the `(ns my.ns)` form.

Eventually I found `executeCodeInNs`, which figures out the namespace of the current file and executes a form in that namespace. 

Would you consider a pull request that removes the feature so the repl feels more "natural"? 
Alternatively, I could strip the `user=>` line from the output, or replace it with the return value of `findNsDeclaration`, but I feel like this would make it even more confusing. Any thoughts on this?

**Comment by jasongilman** [comment]
I'd definitely consider pull requests but I don't want to remove the `executeCodeInNs` feature. It's nice to be able to redefine any function or var in any namespace you jump in without having to switch the REPL to a different namespace. 

## Option A:

I could add a setting to enable or disable execution of code in a namespace but that would be a last resort if we can't think of a better solution.

## Option B:

When you execute the `(ns my.ns)` code it executes this:

``` Clojure
(binding [*ns* (or (find-ns 'my.ns) (find-ns 'user))] 
  (eval (read-string "(ns my.ns)")))
```

When the code is evaluated in the binding it would set the namespace to `my.ns` but binding reverts `*ns*` after it's complete. I was trying to think of a way to detect that an `ns` form had been executed. I came up with the following. It's a lot more complicated but it seems to work.

``` Clojure
(let [ns-set-atom (atom nil)
      orig-in-ns in-ns
      result (with-redefs [in-ns (fn [s]
                                  (reset! ns-set-atom s)
                                  (orig-in-ns s))] 
               (binding [*ns* (or (find-ns 'my.ns) (find-ns 'user))]
                 (eval (read-string "(ns my.ns)"))))]
  (when-let [new-ns @ns-set-atom] 
    (in-ns new-ns))
  result)
```

It replaces the `clojure.core/in-ns` function temporarily to capture whatever namespace that was set in the binding. Then when it's done it will set the namespace outside of the binding. 

## Option C:

Another option would be to get rid of the namespace completely from the REPL display. The REPL would display `=>` without any namespace. Everything that the user executes is either in the namespace defined in the current buffer/file or in `user` if no namespace is specified. This starts to deviate farther from making Proto REPL feel more "natural".

I'm leaning towards Option B because that will achieve what you want without removing the ability to execute the code in a namespace. The only issue is I'm worried about the side effects from temporarily redefining Clojure functions or ways it will break down.

**Comment by qwtel** [comment]
Thanks for the comprehensive answer. Actually, I think option A would be fine :)

Regarding B, I just now understood why the second call to `foo` wouldn't work in my example. So sending the `(ns)` block by itself wasn't doing anything, but for all subsequent calls it would find the ns declaration in the file and use it.

So from my perspective option C seems best now. But to avoid confusion, I think it should show somewhere which namespace the repl will run commands in. Maybe you could include it in the tab title?

**Comment by jasongilman** [comment]
I'm looking at making a change in how the REPL works for issue #24. This would likely address this issue.

**Comment by jasongilman** [comment]
Option C was implemented in 0.9.0.

**Comment by qwtel** [comment]
Just wanted to say I like the current solution 👍
